polar-core/src/kb.rs

```diff
@@ -688,6 +723,125 @@ impl KnowledgeBase {
         errors
     }
 
+    pub fn create_resource_specific_rule_types(&mut self) -> PolarResult<()> {
+        let mut has_relation_rule_types_to_create = HashMap::new();
+
+        // TODO @patrickod refactor RuleTypes & split out
+        // RequiredRuleType struct to record the related
+        // shorthand rule and relation terms.
+
+        // Iterate through all resource block declarations and create
+        // non-required rule types for each relation declaration we observe.
+        //
+        // We create non-required rule types to gracefully account for the case
+        // where users have declared relations ahead of time that are used in
+        // rule or resource definitions.
+        for (object, declarations) in self.resource_blocks.declarations() {
+            for (name, declaration) in declarations.iter() {
+                if let Declaration::Relation(subject) = declaration {
```

Comment:
It would be nice to move this logic requiring intimate knowledge of the inner structure of `ResourceBlocks` into a method on `ResourceBlocks`:

```diff
diff --git a/polar-core/src/kb.rs b/polar-core/src/kb.rs
index 80737271..9799c131 100644
--- a/polar-core/src/kb.rs
+++ b/polar-core/src/kb.rs
@@ -739,13 +739,9 @@ impl KnowledgeBase {
-        for (object, declarations) in self.resource_blocks.declarations() {
-            for (name, declaration) in declarations.iter() {
-                if let Declaration::Relation(subject) = declaration {
-                    has_relation_rule_types_to_create.insert(
-                        (
-                            subject.value().as_symbol()?,
-                            name.value().as_string()?,
-                            object.value().as_symbol()?,
-                        ),
-                        false,
-                    );
-                }
-            }
+        for (subject, name, object) in self.resource_blocks.relation_tuples() {
+            has_relation_rule_types_to_create.insert(
+                (
+                    subject.value().as_symbol()?,
+                    name.value().as_string()?,
+                    object.value().as_symbol()?,
+                ),
+                false,
+            );
diff --git a/polar-core/src/resource_block.rs b/polar-core/src/resource_block.rs
index 66a8e510..ed8670da 100644
--- a/polar-core/src/resource_block.rs
+++ b/polar-core/src/resource_block.rs
@@ -391,0 +392,12 @@ impl ResourceBlocks {
+
+    pub fn relation_tuples(&self) -> Vec<(&Term, &Term, &Term)> {
+        let mut tuples = vec![];
+        for (object, declarations) in self.declarations() {
+            for (name, declaration) in declarations {
+                if let Declaration::Relation(subject) = declaration {
+                    tuples.push((subject, name, object));
+                }
+            }
+        }
+        tuples
+    }
```